### DISCUSSION TITLE
Water Reflections not working with Nature Renderer Trees

### DISCUSSION THREAD

**Discussion by krazyxiii** [visualdesigncafe/nature-renderer#8] [Help]
When I use Nature renderer with my terrain trees the trees are not reflected in the water. When I disable the Nature Renderer script from the terrain the trees are reflected in the water. The same type of tree works just fine also when just added to the scene as a prefab object. Please see screenshots attached. The first Pine tree is a prefab, the next 3 trees are terrain trees using Nature Renderer. Top screenshot Nature Renderer disabled, bottom screenshot Nature Renderer enabled, that's when the trees reflection is not seen. 

Using Unity 2022 LTS, Nature Renderer 2022, Windows 10. Water component is Staggarts Stylized Water 2: 
https://assetstore.unity.com/packages/vfx/shaders/stylized-water-2-170386

Any suggestions on getting this to work would be appreciated.

![image](https://github.com/visualdesigncafe/nature-renderer/assets/73378334/1077d877-ad50-4d61-9993-37db8c0bdd66)

**Comment by NicholasBrown-01** [comment]
Re-read your post, yep it all looks like our issues are related to trees/grass placed via terrain tools.

**Comment by maxplooi** [comment]
Could you check the culling settings that you are using for the reflection camera? It may be that the object's layer is culled for the reflection camera using the camera's [Culling Mask](https://visualdesigncafe.com/nature-renderer/docs/2022/culling#camera-culling-mask). Note that Nature Renderer uses the tree prefab's layer for culling, **not** the terrain's layer.

If that does not work then could you share the settings for the reflection camera? I don't have the Stylized Water asset myself so I can't check the exact setup that they are using for the camera.

**Reply by krazyxiii** [reply]
Hi,

Camera settings, Cullling Mask also  seem to be ok from what I can tell. 
I've attached screenshots. I also tried with the Visual Design Cafe Conifer prefab and same result; no reflection. But if I drop the prefab by just dragging in on to the scene the reflections are fine on the water. I also tried changing the Rendering Path on the Planar Reflection Renderer to Forward+ but that resulted in some kind of exploding lights artifacts on the screen (2nd screen shot attached) 

For the camera I also tried the URP Default Renderer and the Planar Reflections Renderer.

Is there something else I could change in the Nature Global settings to get this to work properly. 

![image](https://github.com/visualdesigncafe/nature-renderer/assets/73378334/b64d1c4d-10de-437c-92d3-f03012453e36)

![image](https://github.com/visualdesigncafe/nature-renderer/assets/73378334/c87dd7a6-7129-4461-ba92-548c2b3086fe)

![image](https://github.com/visualdesigncafe/nature-renderer/assets/73378334/cfea0dbe-a871-43dd-9df6-17baded0f156)

Thanks a lot!

**Comment by maxplooi** [comment]
Thank you for sharing the details. The settings seem to be fine and a similar setup here works fine as well. It must be something specific with the _Stylized Water 2_ asset that is causing the issue. I'll reach out to _Staggart Creations_ to see if I can get a test version of the asset to look into it.

**Comment by krazyxiii** [comment]
Hi Max,

Thanks for the reply and looking into this. I also thought it was something
to do with the Stylized Water asset but those trees, which were placed
using the terrain tree paint tool do show up reflected in the water when I
disable the Nature Renderer component on the Terrain. If I place the same
tree prefab manually in front of the water the reflection works just fine.
Anyway I hope that you can work with the Staggart's Creation developer to
see if this can be resolved. The 2 assets work great together in my
opinion, both those assets along with Cozy Weather 3 are very important
assets in my project. Please keep me posted once you've talked to him and
come up with a plan.

Thank you again!

Ivan

**Comment by maxplooi** [comment]
I looked into it for you after contacting Staggart Creations and the main issue is that the Stylized Water asset is using a deprecated rendering method from URP. While the deprecated method works fine for rendering (and there doesn't seem to be a good alternative in URP), it ignores all the rendering callbacks from URP which are needed by Nature Renderer to register the camera and render the detail objects. 

For now, the only workaround is to manually register and render the camera when rendering reflections. You can do this by adding the following code to the **PlanarReflectionRenderer.cs** script of Stylized Water around line 286 (just before `UniversalRenderPipeline.RenderSingleCamera(context, m_reflectionCamera)`):

```c#
using VisualDesignCafe.Rendering.Instancing;

// Register the reflection camera for Nature Renderer
var cameraId = RendererPool.RegisterCamera( m_reflectionCamera );

// Render the instanced objects (details and trees)
RendererPool.GetCamera( cameraId ).Render();
``` 

Example:
![stylized-water-code-example](https://github.com/visualdesigncafe/nature-renderer/assets/9071464/0debaedd-e09a-4e5b-921c-b6854a822a54)

**Comment by krazyxiii** [comment]
Hi,

I don't know how I missed this reply, but I will look into your workaround
and let you know if it works or not, without breaking anything else. If for
some reason it's not an ideal solution then I will just place prefabs in
the areas of the terrain where I want them to be reflected in the water,or
look into using a different water shader.

Thanks again!

Ivan
